syntax = "proto3";

/*
Week 12 â€” gRPC calculator service

This file defines a small calculator service used to demonstrate:
  - unary RPCs
  - typed messages
  - basic error handling

To regenerate Python stubs:
  make proto-gen
*/

package calculator;

// Calculator service with basic operations.
service Calculator {
  rpc Add (CalcRequest) returns (CalcResponse);
  rpc Subtract (CalcRequest) returns (CalcResponse);
  rpc Multiply (CalcRequest) returns (CalcResponse);
  rpc Divide (CalcRequest) returns (CalcResponse);

  rpc Echo (EchoRequest) returns (EchoResponse);
  rpc Sha256Hash (HashRequest) returns (HashResponse);

  rpc GetStats (EmptyRequest) returns (StatsResponse);
}

// Basic calculator request.
message CalcRequest {
  double a = 1;
  double b = 2;
}

// Basic calculator response.
message CalcResponse {
  double result = 1;
  string operation = 2;
  int64 timestamp = 3;
}

// Echo request/response.
message EchoRequest {
  string message = 1;
}

message EchoResponse {
  string message = 1;
  int64 timestamp = 2;
}

// Hash request/response.
message HashRequest {
  string input = 1;
}

message HashResponse {
  string hash = 1;
  int64 timestamp = 2;
}

// Empty request.
message EmptyRequest {}

// Statistics response.
message StatsResponse {
  map<string, int32> call_counts = 1;
  int32 total_calls = 2;
  int64 uptime_seconds = 3;
}
